packages/hash/api/src/model/knowledge/link.model.ts

```diff
@@ -111,19 +110,191 @@ export default class {
+    const hasOrderedSiblingLink = siblingLinks[0]?.index !== undefined;
+
+    const index = hasOrderedSiblingLink
+      ? // if siblings are ordered and an index is provided, use the provided index
+        params.index ??
+        // if siblings are ordered and no index is provided, default to the end of the list of links
+        siblingLinks.length
+      : siblingLinks.length === 0
+      ? // if siblings are not ordered because there are no siblings, allow the link to be ordered
+        params.index
+      : // if siblings are not ordered and there are siblings, don't allow the link to be ordered
+        undefined;
+
+    if (index !== undefined) {
+      /** @todo: rely on Graph API to validate the index */
+      if (index < 0 || index > siblingLinks.length) {
+        throw new Error("Provided link index is out of bounds");
+      }
+
+      await Promise.all(
+        siblingLinks
+          .filter((sibling) => sibling.index! >= index)
+          .map((sibling) =>
+            sibling.updateWithoutUpdatingSiblings(graphApi, {
+              updatedIndex: sibling.index! + 1,
+              updatedBy: createdBy,
+            }),
+          ),
+      );
+    }
+
+    return await LinkModel.createLinkWithoutUpdatingSiblings(graphApi, {
+      ...params,
+      index,
+    });
+  }
+
+  private async updateWithoutUpdatingSiblings(
+    graphApi: GraphApi,
+    params: { updatedIndex: number; updatedBy: string },
+  ) {
+    const { updatedIndex, updatedBy } = params;
+
+    const { index: previousIndex } = this;
+
+    if (previousIndex === undefined) {
+      throw new Error("Cannot make an un-ordered link ordered");
+    }
+
+    if (previousIndex === updatedIndex) {
+      throw new Error("No-op: link already has this index");
+    }
+
+    /**
+     * @todo: call dedicated Graph API method to update the index of a link instead of re-creating the link manually
+     * @see https://app.asana.com/0/1202805690238892/1203031430417465/f
+     */
+    await this.removeWithoutUpdatingSiblings(graphApi, {
+      removedBy: updatedBy,
+    });
+
+    const updatedLink = await LinkModel.createLinkWithoutUpdatingSiblings(
+      graphApi,
+      {
+        ...this,
+        index: updatedIndex,
+        createdBy: updatedBy,
+      },
+    );
+
+    return updatedLink;
+  }
+
+  /**
+   * Update the link
+   *
+   * @param params.updatedIndex - the updated index of the link
+   * @param params.updatedBy - the account updating the link
+   */
+  async update(
+    graphApi: GraphApi,
+    params: { updatedIndex: number; updatedBy: string },
+  ) {
+    const { updatedIndex, updatedBy } = params;
+
+    const { index: previousIndex, linkTypeModel } = this;
+
+    const updatedLink = await this.updateWithoutUpdatingSiblings(graphApi, {
+      updatedIndex,
+      updatedBy,
+    });
+
+    const siblingLinks = await this.sourceEntityModel.getOutgoingLinks(
+      graphApi,
+      { linkTypeModel },
+    );
+
+    if (previousIndex === undefined) {
```

Comment:
Don't we want to do this before calling `this.updateWithoutUpdatingSiblings`?

Reply:
Yes, and [I do](https://github.com/hashintel/hash/blob/8b8bcd28e5e8283807753fd2c5a40ee57e3a85e9/packages/hash/api/src/model/knowledge/link.model.ts#L244-L247). This is to get around the type error (`this.index` can be `undefined` otherwise). Would you like me to cast instead?

Reply:
Makes sense, I'm not sure we need it duplicated. Is there a way we can change the signatures to let TS know?

If not then I'd still move the block I've commented on up about 10 lines. It's just quite strange to check if it's undefined _after_ calling things with it

Reply:
Moved the check in https://github.com/hashintel/hash/pull/1081/commits/64053b726e58d4d39ee11ee49fe13208ea608786